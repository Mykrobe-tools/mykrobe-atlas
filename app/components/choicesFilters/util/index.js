/* @flow */

import _ from 'lodash';

import type { Choice } from '../types';

export const titleForChoice = (choice: Choice, delimeter: string = ' › ') => {
  return choice.titles
    ? choice.titles.join(delimeter)
    : choice.title || 'Untitled';
};

// build an inverse tree of titles

export const choiceTitleTree = _.memoize(
  choices => {
    let titleTree = {};
    const choicesKeys = Object.keys(choices);
    choicesKeys.forEach(key => {
      const choice = choices[key];
      let node = titleTree;
      if (choice.titles) {
        const reverseTitles = [...choice.titles].reverse();
        reverseTitles.forEach(title => {
          if (!node[title]) {
            node[title] = {};
          }
          node = node[title];
        });
      }
    });
    return titleTree;
  },
  choices => JSON.stringify(choices)
);

export const shortesTitleForChoiceWithKeyInChoices = (
  choiceKey: string,
  choices: { [string]: Choice },
  delimeter: string = ' › '
) => {
  const titleTree = choiceTitleTree(choices);
  // traverse and stop when we are the only entry
  const choice = choices[choiceKey];
  const reverseTitles = [...choice.titles].reverse();
  let titles = [];
  let node = titleTree;
  let i = 0;
  while (i < reverseTitles.length) {
    const title = reverseTitles[i];
    titles.push(title);
    node = node[title];
    if (!node) {
      break;
    }
    if (Object.keys(node).length === 1) {
      break;
    }
    i++;
  }
  titles.reverse();
  return titles.join(delimeter);
};
