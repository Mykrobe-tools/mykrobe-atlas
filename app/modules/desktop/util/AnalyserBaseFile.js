/* @flow */

import EventEmitter from 'events';

// TODO: we can remove the transformer - it is created on demand by a redux selector
import AnalyserJsonTransformer from '../../experiments/util/AnalyserJsonTransformer';
import * as APIConstants from '../../../constants/APIConstants';

class AnalyserBaseFile extends EventEmitter {
  extensionForFileName(fileName: string) {
    const extension = fileName.substr(fileName.lastIndexOf('.'));
    return extension.toLowerCase();
  }

  analyseFile(filePaths: Array<string>, id: string = ''): AnalyserBaseFile {
    this.cancel();
    for (let i = 0; i < filePaths.length; i++) {
      const filePath = filePaths[i];
      const extension = this.extensionForFileName(filePath);
      if (extension === '.json') {
        return this.analyseJsonFile(filePath);
      }
      if (
        !APIConstants.API_SAMPLE_EXTENSIONS_ARRAY_WITH_DOTS.includes(extension)
      ) {
        const acceptable = APIConstants.API_SAMPLE_EXTENSIONS_ARRAY_WITH_DOTS.join(
          ', '
        );
        this.failWithError(
          `Can only process files with extension: ${acceptable} - not ${extension}`
        );
        return this;
      }
      if (filePath.indexOf(' ') > 0) {
        this.failWithError(
          'Mykrobe does not currently work with files or paths containing spaces'
        );
        return this;
      }
    }
    return this.analyseBinaryFile(filePaths, id);
  }

  failWithError(err: string | Error) {
    let message = err;
    if (err.message) {
      message = err.message;
    }
    setTimeout(() => {
      this.emit('error', {
        description: message,
      });
    }, 0);
  }

  doneWithJsonString(jsonString: string) {
    // const fs = require('fs');
    // fs.writeFileSync('doneWithJsonString.json', jsonString);
    const transformer = new AnalyserJsonTransformer();
    transformer
      .transform(jsonString)
      .then(result => {
        // const { json, transformed } = result;
        // console.log('json', json);
        // console.log('transformed', transformed);
        this.emit('done', result);
      })
      .catch(err => {
        this.failWithError(err);
      });
  }

  analyseJsonFile(file: File | string): AnalyserBaseFile {
    console.log('analyseJsonFile', file);
    return this;
  }

  analyseBinaryFile(filePaths: Array<string>): AnalyserBaseFile {
    console.log('analyseBinaryFile', filePaths);
    return this;
  }

  analyseRemoteFile(file: File | string): AnalyserBaseFile {
    console.log('analyseRemoteFile', file);
    return this;
  }

  cancel(): void {}
}

export default AnalyserBaseFile;
