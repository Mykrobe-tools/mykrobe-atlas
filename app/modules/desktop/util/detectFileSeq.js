/* @flow */

const fs = require('fs');
const path = require('path');

import { isString, isArray } from 'makeandship-js-common/src/utils/is';

import { API_SAMPLE_EXTENSIONS_ARRAY } from '../../../constants/APIConstants';

export const detectFileSeqForFileNameInDir = (
  fileName: string,
  dir: Array<string>
) => {
  if (!isString(fileName)) {
    return;
  }
  if (!isArray(dir)) {
    return;
  }

  let elements = fileName.split('.');
  if (elements.length === 0) {
    return;
  }

  // find the last element ending with a digit which directly preceeds a valid extension
  let searching = true;
  let extensions = [];
  let element;

  // pop elements that are valid extensions
  while (searching && elements.length > 0) {
    element = elements.pop();
    if (API_SAMPLE_EXTENSIONS_ARRAY.includes(element.toLowerCase())) {
      extensions.splice(0, 0, element);
    } else {
      searching = false;
    }
  }

  if (searching || !element || !isString(element)) {
    return;
  }

  // check if next element ends in digit

  const matches = element.match(/(.+)(\d+)$/);

  if (matches && matches.length === 3) {
    const elementPrefix = matches[1];
    const elementNumber = parseInt(matches[2]);

    // recreate filename +/- 1
    const elementBefore = `${elementPrefix}${elementNumber - 1}`;
    const elementAfter = `${elementPrefix}${elementNumber + 1}`;
    const elementsBefore = [...elements, elementBefore, ...extensions];
    const elementsAfter = [...elements, elementAfter, ...extensions];
    const filenameBefore = elementsBefore.join('.');
    const filenameAfter = elementsAfter.join('.');

    if (dir.includes(filenameBefore)) {
      return filenameBefore;
    } else if (dir.includes(filenameAfter)) {
      return filenameAfter;
    }
  }
};

export const detectFileSeqForFilePathInDir = (
  filePath: string,
  dir: Array<string>
) => {
  const seqName = detectFileSeqForFileNameInDir(path.basename(filePath), dir);
  if (seqName) {
    return path.join(path.dirname(filePath), seqName);
  }
};

const detectFileSeq = (filePath: string) => {
  const dir = fs.readdirSync(path.dirname(filePath));
  return detectFileSeqForFilePathInDir(filePath, dir);
};

export default detectFileSeq;
