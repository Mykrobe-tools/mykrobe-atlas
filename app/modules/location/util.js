/* @flow */

// rather than use query-string, these functions preserve order of keys
// this is necessary when dynamically adding and removing filters
// so that the ui will maintain order

export const parseQuery = (query: string): any => {
  const searchParams = new URLSearchParams(query);
  const parsed = {};
  for (let param of searchParams) {
    const key = param[0];
    const value = param[1];
    parsed[key] = value;
  }
  return parsed;
};

export const stringifyQuery = (query: any): string => {
  var searchParams = new URLSearchParams('');
  const queryKeys = Object.keys(query);
  for (let i = 0; i < queryKeys.length; i++) {
    const key = queryKeys[i];
    const value = query[key];
    searchParams.set(key, value);
  }
  return searchParams.toString();
};
