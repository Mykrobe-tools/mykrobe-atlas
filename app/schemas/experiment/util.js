/* @flow */

import _ from 'lodash';

// TODO: move into common library

export const keysForSchema = (
  schema: any,
  subSchema?: any,
  currentKeyPath: Array<string> = []
): Array<string> => {
  let keys = [];
  if (!subSchema) {
    subSchema = schema;
  }
  if (subSchema.properties) {
    Object.entries(subSchema.properties).forEach(([key, value]) => {
      const keyPath = currentKeyPath.concat(key);
      if (typeof value === 'object' && value !== null) {
        // don't add this key path unless it can represent a value
        if (value.type && value.type !== 'object') {
          keys.push(keyPath.join('.'));
        }
        keys = keys.concat(keysForSchema(schema, value, keyPath));
      }
    });
  } else if (subSchema.$ref) {
    const definition = findSchemaDefinitionForRef(schema, subSchema['$ref']);
    keys = keys.concat(keysForSchema(schema, definition, currentKeyPath));
  }
  return keys;
};

export const findSchemaDefinitionForRef = (schema: any, ref: string) => {
  const parts = ref.split('/').slice(1);
  const definition = _.get(schema, parts);
  return definition;
};

// TODO: this is currently very naive, e.g. will not skip key paths due to visibility rules

export const completenessForSchemaAndData = (schema: any, data: any) => {
  let complete = 0;
  const keys = keysForSchema(schema);
  const total = keys.length;
  keys.forEach(key => {
    const value = _.get(data, key);
    if (value) {
      complete++;
    }
  });
  return { complete, total };
};
