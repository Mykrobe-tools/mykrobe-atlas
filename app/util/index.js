/* @flow */

export const notImplemented = () => {
  alert('This feature is not yet implemented');
};
