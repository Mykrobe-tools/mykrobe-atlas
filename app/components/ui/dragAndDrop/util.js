/* @flow */

import { extensionForFileName } from '../../../util';

export const shouldAcceptDropEvent = (e: any) => {
  const dt = e.dataTransfer;
  if (
    !(
      dt.types &&
      (dt.types.indexOf
        ? dt.types.indexOf('Files') !== -1
        : dt.types.contains('Files'))
    )
  ) {
    return false;
  }
  return true;
};

export const shouldAcceptDropEventForExtensions = (
  e: any,
  accept: Array<*>
) => {
  if (!shouldAcceptDropEvent(e)) {
    return false;
  }
  for (let i = 0; i < e.dataTransfer.files.length; i++) {
    const file = e.dataTransfer.files[i];
    const extension = extensionForFileName(file.name);
    if (!accept.includes(extension)) {
      return false;
    }
  }
  return true;
};
