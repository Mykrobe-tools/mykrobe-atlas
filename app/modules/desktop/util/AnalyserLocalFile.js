/* @flow */

import EventEmitter from 'events';
import path from 'path';
import { spawn } from 'child_process';
import readline from 'readline';
import fs from 'fs-extra';

import { isString } from 'makeandship-js-common/src/utils/is';

import AnalyserJsonTransformer from '../../experiments/util/AnalyserJsonTransformer';
import * as APIConstants from '../../../constants/APIConstants';

import { pathToBin } from './pathToBin';
import { extensionForFileName } from '../../../util';
import isAnalyserError from './isAnalyserError';

const DEBUG =
  process.env.DEBUG_PRODUCTION === '1' ||
  process.env.NODE_ENV === 'development';

const tmp = require('tmp');
const app = require('electron').remote.app;

class AnalyserLocalFile extends EventEmitter {
  jsonBuffer: string;
  isBufferingJson: boolean;
  processExited: boolean;
  child: child_process$ChildProcess;
  didReceiveError: boolean;
  readLineExited: boolean;
  tmpObj: ?Object;

  analyseFile(filePaths: Array<string>, id: string = ''): AnalyserLocalFile {
    this.cancel();
    for (let i = 0; i < filePaths.length; i++) {
      const filePath = filePaths[i];
      const extension = extensionForFileName(filePath);
      if (extension === '.json') {
        return this.analyseJsonFile(filePath);
      }
      if (
        !APIConstants.API_SAMPLE_EXTENSIONS_ARRAY_WITH_DOTS.includes(extension)
      ) {
        const acceptable = APIConstants.API_SAMPLE_EXTENSIONS_ARRAY_WITH_DOTS.join(
          ', '
        );
        this.failWithError(
          `Mykrobe can only process files with extension: ${acceptable} - not ${extension}`
        );
        return this;
      }
    }
    return this.analyseBinaryFile(filePaths, id);
  }

  failWithError(err: string | Error) {
    if (!DEBUG) {
      // fail immediately
      this.cancel();
    }
    let message = err;
    if (err.message) {
      message = err.message;
    }
    setTimeout(() => {
      this.emit('error', {
        description: message,
      });
    }, 0);
  }

  doneWithJsonString(jsonString: string) {
    this.cleanup();
    const transformer = new AnalyserJsonTransformer();
    transformer
      .transform(jsonString)
      .then((result) => {
        // const { json, transformed } = result;
        // console.log('json', json);
        // console.log('transformed', transformed);
        this.emit('done', result);
      })
      .catch((err) => {
        this.failWithError(err);
      });
  }

  constructor() {
    super();
    app &&
      app.on('quit', () => {
        this.cancel();
        console.log('ev:app quit');
      });
  }

  analyseJsonFile = (file: File | string): AnalyserLocalFile => {
    if (isString(file)) {
      fs.readFile(file, (err, data) => {
        if (err) throw err;
        const dataString = data.toString();
        this.doneWithJsonString(dataString);
      });
    } else {
      const reader = new FileReader();
      reader.onload = () => {
        const dataString = reader.result;
        this.doneWithJsonString(dataString);
      };

      reader.onerror = (e) => {
        this.failWithError(
          `FileReader failed with error code ${e.target.error.code}`
        );
      };

      reader.readAsText(file);
    }
    return this;
  };

  analyseBinaryFile = (filePaths: Array<string>): AnalyserLocalFile => {
    this.tmpObj = tmp.dirSync({ prefix: 'mykrobe-' });

    const skeletonDir = path.join(this.tmpObj.name, 'skeleton');
    const tmpDir = path.join(this.tmpObj.name, 'tmp');

    fs.ensureDirSync(skeletonDir);
    fs.ensureDirSync(tmpDir);

    // Sample name should not contain whitespace, replace with '-'
    let sampleName = path.parse(filePaths[0]).name;
    sampleName = sampleName.replace(/\s+/g, '-');

    const pathToBinValue = pathToBin();

    // FIXME: if we don't use --force, occasionally get error - some files cached??

    const args = [
      'predict',
      '--force',
      sampleName,
      'tb',
      filePaths.length > 1 ? '--seq' : '-1',
      ...filePaths,
      '--tmp',
      tmpDir,
      '--skeleton_dir',
      skeletonDir,
      '--format',
      'json',
      '--guess_sequence_method',
      '--quiet',
    ];

    console.log('Guess at command line:', pathToBinValue, args.join(' '));
    console.log('Spawning executable at path:', pathToBinValue);
    console.log('With arguments:', args);
    const child = spawn(pathToBinValue, args);
    this.setChildProcess(child);

    if (DEBUG) {
      const tmpDir = tmp.dirSync({ prefix: 'mykrobe-debug-' }).name;
      const cmdPath = path.join(tmpDir, 'AnalyserLocalFile.cmd.txt');
      console.log('Logging command to', cmdPath);
      fs.writeFileSync(cmdPath, `${pathToBinValue} ${args.join(' ')}`);
    }

    return this.monitorChildProcess();
  };

  setChildProcess = (child: child_process$ChildProcess) => {
    this.child = child;
    return this;
  };

  monitorChildProcess = () => {
    this.jsonBuffer = '';
    this.isBufferingJson = false;
    this.processExited = false;
    this.readLineExited = false;

    if (!this.child) {
      this.failWithError('Failed to start child process');
      return this;
    }

    this.child.on('error', (err) => {
      console.log('Failed to start child process.', err);
      this.failWithError(err);
    });

    if (DEBUG) {
      const tmpDir = tmp.dirSync({ prefix: 'mykrobe-debug-' }).name;
      const stdoutPath = path.join(tmpDir, 'AnalyserLocalFile.stdout.txt');
      const stderrPath = path.join(tmpDir, 'AnalyserLocalFile.stderr.txt');
      console.log('Logging stdout and stderr to', stdoutPath, stderrPath);
      const stdoutStream = fs.createWriteStream(stdoutPath);
      const stderrStream = fs.createWriteStream(stderrPath);
      this.child.stdout.pipe(stdoutStream);
      this.child.stderr.pipe(stderrStream);
    }

    readline
      .createInterface({
        input: this.child.stdout,
      })
      .on('line', (line) => {
        if (this.didReceiveError) {
          return;
        }
        if (!this.isBufferingJson) {
          console.log(line);
        } else {
          console.log('Received json, muted');
        }
        if (line.indexOf('Progress:') >= 0) {
          console.log('progress');
          // we get a string like "[15 Oct 2017 16:19:47-Kac] Progress: 130,000/454,797"
          // extract groups of digits
          const trimmed = line.substr(line.indexOf('Progress:'));
          const digitGroups = trimmed.replace(/,/g, '').match(/\d+/g);
          if (digitGroups.length > 1) {
            const progress = parseInt(digitGroups[0]);
            const total = parseInt(digitGroups[1]);
            console.log('progress:' + progress);
            console.log('total:' + total);
            this.emit('progress', {
              progress,
              total,
            });
          }
        }

        if (this.isBufferingJson) {
          this.jsonBuffer += line;
        } else if (line.indexOf('{') !== -1) {
          // start collecting as soon as we see { in the buffer
          this.isBufferingJson = true;
          this.jsonBuffer = line.substr(line.indexOf('{'));
        }
      })
      .on('close', () => {
        console.log('readline stdin close');
        this.onReadLineExited();
      });

    readline
      .createInterface({
        input: this.child.stderr,
      })
      .on('line', (line) => {
        if (isAnalyserError(line)) {
          this.onReadLineError(line);
        } else {
          console.log('IGNORING ERROR: ' + line);
        }
      })
      .on('close', () => {
        console.log('readline stderr close');
      });

    this.child.on('exit', (code) => {
      console.log('Processing exited with code: ' + code);
      // 0 = exited normally, null = cancelled
      if (code === 0 || code === null) {
        this.onProcessExited();
      } else {
        this.failWithError(`Process exit unexpectedly with code ${code}`);
      }
    });

    return this;
  };

  onProcessExited = () => {
    this.processExited = true;
    this.checkDone();
  };

  onReadLineExited = () => {
    this.readLineExited = true;
    this.checkDone();
  };

  checkDone = () => {
    // wait until both readline and process have exited;
    // this can happen in different orders on different systems
    if (this.readLineExited && this.processExited && this.jsonBuffer.length) {
      this.doneWithJsonString(this.jsonBuffer);
    }
  };

  onReadLineError = (line: string) => {
    console.log('ERROR: ' + line);
    this.didReceiveError = true;
    this.failWithError(line);
  };

  cancel() {
    if (this.child) {
      this.child.kill();
      delete this.child;
    }
    this.cleanup();
  }

  cleanup() {
    console.log('cleanup');
    // this.tmpObj.removeCallback() doesn't always work
    this.tmpObj && fs.removeSync(this.tmpObj.name);
  }
}

export default AnalyserLocalFile;
