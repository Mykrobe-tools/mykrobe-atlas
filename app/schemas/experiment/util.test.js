/* @flow */

import { experimentSchema } from './index';
import { keysForSchema } from './util';

describe('experiment schema util', () => {
  it('should derive keys as expected', async () => {
    expect(keysForSchema(experimentSchema)).toMatchSnapshot();
  });
});
