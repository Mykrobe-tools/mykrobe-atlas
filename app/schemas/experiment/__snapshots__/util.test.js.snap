// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`experiment schema util should derive keys as expected 1`] = `
Array [
  "sampleId",
  "leafId",
  "files",
  "metadata.sample.labId",
  "metadata.sample.isolateId",
  "metadata.sample.collectionDate",
  "metadata.sample.prospectiveIsolate",
  "metadata.sample.countryIsolate",
  "metadata.sample.cityIsolate",
  "metadata.sample.longitudeIsolate",
  "metadata.sample.latitudeIsolate",
  "metadata.sample.dateArrived",
  "metadata.sample.anatomicalOrigin",
  "metadata.sample.smear",
  "metadata.genotyping.wgsPlatform",
  "metadata.genotyping.wgsPlatformOther",
  "metadata.genotyping.otherGenotypeInformation",
  "metadata.genotyping.genexpert",
  "metadata.genotyping.hain",
  "metadata.genotyping.hainRif",
  "metadata.genotyping.hainInh",
  "metadata.genotyping.hainFl",
  "metadata.genotyping.hainAm",
  "metadata.genotyping.hainEth",
  "metadata.phenotyping.phenotypeInformationFirstLineDrugs",
  "metadata.phenotyping.rifampicin.susceptibility",
  "metadata.phenotyping.rifampicin.method",
  "metadata.phenotyping.ethambutol.susceptibility",
  "metadata.phenotyping.ethambutol.method",
  "metadata.phenotyping.pyrazinamide.susceptibility",
  "metadata.phenotyping.pyrazinamide.method",
  "metadata.phenotyping.isoniazid.susceptibility",
  "metadata.phenotyping.isoniazid.method",
  "metadata.phenotyping.phenotypeInformationOtherDrugs",
  "metadata.phenotyping.rifabutin.susceptibility",
  "metadata.phenotyping.rifabutin.method",
  "metadata.phenotyping.ofloxacin.susceptibility",
  "metadata.phenotyping.ofloxacin.method",
  "metadata.phenotyping.ciprofloxacin.susceptibility",
  "metadata.phenotyping.ciprofloxacin.method",
  "metadata.phenotyping.levofloxacin.susceptibility",
  "metadata.phenotyping.levofloxacin.method",
  "metadata.phenotyping.gatifloxacin.susceptibility",
  "metadata.phenotyping.gatifloxacin.method",
  "metadata.phenotyping.amikacin.susceptibility",
  "metadata.phenotyping.amikacin.method",
  "metadata.phenotyping.kanamycin.susceptibility",
  "metadata.phenotyping.kanamycin.method",
  "metadata.phenotyping.gentamicin.susceptibility",
  "metadata.phenotyping.gentamicin.method",
  "metadata.phenotyping.streptomycin.susceptibility",
  "metadata.phenotyping.streptomycin.method",
  "metadata.phenotyping.capreomycin.susceptibility",
  "metadata.phenotyping.capreomycin.method",
  "metadata.phenotyping.clofazimine.susceptibility",
  "metadata.phenotyping.clofazimine.method",
  "metadata.phenotyping.pas.susceptibility",
  "metadata.phenotyping.pas.method",
  "metadata.phenotyping.linezolid.susceptibility",
  "metadata.phenotyping.linezolid.method",
  "metadata.phenotyping.ethionamide.susceptibility",
  "metadata.phenotyping.ethionamide.method",
  "metadata.phenotyping.terizidone.susceptibility",
  "metadata.phenotyping.terizidone.method",
  "metadata.phenotyping.amoxicilinClavulanate.susceptibility",
  "metadata.phenotyping.amoxicilinClavulanate.method",
  "metadata.phenotyping.thioacetazone.susceptibility",
  "metadata.phenotyping.thioacetazone.method",
  "metadata.phenotyping.imipenem.susceptibility",
  "metadata.phenotyping.imipenem.method",
  "metadata.phenotyping.meropenem.susceptibility",
  "metadata.phenotyping.meropenem.method",
  "metadata.phenotyping.clarythromycin.susceptibility",
  "metadata.phenotyping.clarythromycin.method",
  "metadata.phenotyping.highDoseIsoniazid.susceptibility",
  "metadata.phenotyping.highDoseIsoniazid.method",
  "metadata.phenotyping.bedaquiline.susceptibility",
  "metadata.phenotyping.bedaquiline.method",
  "metadata.phenotyping.delamanid.susceptibility",
  "metadata.phenotyping.delamanid.method",
  "metadata.phenotyping.prothionamide.susceptibility",
  "metadata.phenotyping.prothionamide.method",
  "metadata.phenotyping.pretomanid.susceptibility",
  "metadata.phenotyping.pretomanid.method",
  "metadata.treatment.previousTbInformation",
  "metadata.treatment.recentMdrTb",
  "metadata.treatment.priorTreatmentDate",
  "metadata.treatment.tbProphylaxis",
  "metadata.treatment.tbProphylaxisDate",
  "metadata.treatment.currentTbInformation",
  "metadata.treatment.startProgrammaticTreatment",
  "metadata.treatment.intensiveStartDate",
  "metadata.treatment.intensiveStopDate",
  "metadata.treatment.startProgrammaticContinuationTreatment",
  "metadata.treatment.continuation.start",
  "metadata.treatment.continuation.stop",
  "metadata.treatment.nonStandardTreatment",
  "metadata.treatment.outsideStandardPhaseRifampicinRifabutin.start",
  "metadata.treatment.outsideStandardPhaseRifampicinRifabutin.stop",
  "metadata.treatment.outsideStandardPhaseEthambutol.start",
  "metadata.treatment.outsideStandardPhaseEthambutol.stop",
  "metadata.treatment.outsideStandardPhasePyrazinamide.start",
  "metadata.treatment.outsideStandardPhasePyrazinamide.stop",
  "metadata.treatment.outsideStandardPhaseIsoniazid.start",
  "metadata.treatment.outsideStandardPhaseIsoniazid.stop",
  "metadata.treatment.outsideStandardPhaseOfloxacin.start",
  "metadata.treatment.outsideStandardPhaseOfloxacin.stop",
  "metadata.treatment.outsideStandardPhaseMoxifloxacin.start",
  "metadata.treatment.outsideStandardPhaseMoxifloxacin.stop",
  "metadata.treatment.outsideStandardPhaseLevofloxacin.start",
  "metadata.treatment.outsideStandardPhaseLevofloxacin.stop",
  "metadata.treatment.outsideStandardPhaseGatifloxacin.start",
  "metadata.treatment.outsideStandardPhaseGatifloxacin.stop",
  "metadata.treatment.outsideStandardPhaseAmikacin.start",
  "metadata.treatment.outsideStandardPhaseAmikacin.stop",
  "metadata.treatment.outsideStandardPhaseGentamicin.start",
  "metadata.treatment.outsideStandardPhaseGentamicin.stop",
  "metadata.treatment.outsideStandardPhaseStreptomycin.start",
  "metadata.treatment.outsideStandardPhaseStreptomycin.stop",
  "metadata.treatment.outsideStandardPhaseCapreomycin.start",
  "metadata.treatment.outsideStandardPhaseCapreomycin.stop",
  "metadata.treatment.outsideStandardPhaseClofazimine.start",
  "metadata.treatment.outsideStandardPhaseClofazimine.stop",
  "metadata.treatment.outsideStandardPhasePas.start",
  "metadata.treatment.outsideStandardPhasePas.stop",
  "metadata.treatment.outsideStandardPhaseLinezolid.start",
  "metadata.treatment.outsideStandardPhaseLinezolid.stop",
  "metadata.treatment.outsideStandardPhaseEthionamideProthionamide.start",
  "metadata.treatment.outsideStandardPhaseEthionamideProthionamide.stop",
  "metadata.treatment.outsideStandardPhaseTerizidone.start",
  "metadata.treatment.outsideStandardPhaseTerizidone.stop",
  "metadata.treatment.outsideStandardPhaseAmoxicilinClavulanate.start",
  "metadata.treatment.outsideStandardPhaseAmoxicilinClavulanate.stop",
  "metadata.treatment.outsideStandardPhaseThioacetazone.start",
  "metadata.treatment.outsideStandardPhaseThioacetazone.stop",
  "metadata.treatment.outsideStandardPhaseImipenemImipenemcilastatin.start",
  "metadata.treatment.outsideStandardPhaseImipenemImipenemcilastatin.stop",
  "metadata.treatment.outsideStandardPhaseMeropenem.start",
  "metadata.treatment.outsideStandardPhaseMeropenem.stop",
  "metadata.treatment.outsideStandardPhaseClarythromycin.start",
  "metadata.treatment.outsideStandardPhaseClarythromycin.stop",
  "metadata.treatment.outsideStandardPhaseHighDoseIsoniazid.start",
  "metadata.treatment.outsideStandardPhaseHighDoseIsoniazid.stop",
  "metadata.patient.patientId",
  "metadata.patient.siteId",
  "metadata.patient.genderAtBirth",
  "metadata.patient.countryOfBirth",
  "metadata.patient.age",
  "metadata.patient.bmi",
  "metadata.patient.injectingDrugUse",
  "metadata.patient.homeless",
  "metadata.patient.imprisoned",
  "metadata.patient.smoker",
  "metadata.patient.diabetic",
  "metadata.patient.hivStatus",
  "metadata.patient.art",
  "metadata.outcome.sputumSmearConversion",
  "metadata.outcome.sputumCultureConversion",
  "metadata.outcome.whoOutcomeCategory",
  "metadata.outcome.dateOfDeath",
  "results",
  "awaitingFirstDistanceResult",
]
`;

exports[`experiment schema util should determine completeness as expected 1`] = `
Object {
  "complete": 2,
  "total": 162,
}
`;
